golem-test-framework/src/config/mod.rs

```diff
@@ -0,0 +1,57 @@
+// Copyright 2024 Golem Cloud
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::path::PathBuf;
+use std::sync::Arc;
+
+pub use cli::{CliParams, CliTestDependencies};
+pub use env::EnvBasedTestDependencies;
+
+use crate::components::rdb::Rdb;
+use crate::components::redis::Redis;
+use crate::components::redis_monitor::RedisMonitor;
+use crate::components::shard_manager::ShardManager;
+use crate::components::template_service::TemplateService;
+use crate::components::worker_executor_cluster::WorkerExecutorCluster;
+use crate::components::worker_service::WorkerService;
+
+mod cli;
+mod env;
+
+pub trait TestDependencies {
+    fn rdb(&self) -> Arc<dyn Rdb + Send + Sync + 'static>;
+    fn redis(&self) -> Arc<dyn Redis + Send + Sync + 'static>;
+    fn redis_monitor(&self) -> Arc<dyn RedisMonitor + Send + Sync + 'static>;
+    fn shard_manager(&self) -> Arc<dyn ShardManager + Send + Sync + 'static>;
+    fn template_directory(&self) -> PathBuf;
+    fn template_service(&self) -> Arc<dyn TemplateService + Send + Sync + 'static>;
+    fn worker_service(&self) -> Arc<dyn WorkerService + Send + Sync + 'static>;
+    fn worker_executor_cluster(&self) -> Arc<dyn WorkerExecutorCluster + Send + Sync + 'static>;
+
+    fn kill_all(&self) {
```

Comment:
Methods here return `Arc` for real services. And I believe this is the reason why we have to use custom `kill/kill_all` methods instead of `Drop`. This is the reason why I created `*Info` structs. Most of the tests need only `*Info` which implements `Clon` and has no `Drop`. The only tests that needs real services is `sharding`, but in `sharding` we don't need `Arc` either because there is a single thread that can own real services with `Drop`.

The approach with `kill`/`kill_all` as a bit risky because someone can add a field with `Drop` in some of the structs and incorrectly assume it should work.

Reply:
I was thinking of keeping your approach with the `*Info` descriptors but in the end felt like this one is safer/easier to understand for me. The reason is that those are essentially weak references to the running services (not real references but host/port information) completely decoupled from the lifetime of the started services, while with my approach they are real shared reference counted references to the thing that controls the running services. This way it is guaranteed that if you have a reference to one of these dependencies, it still exists and that anyone who has a reference for it can "control it" (for example restart). On the other hand it is true that with an explicit `kill` the guarantee that "it still runs" is not really true anymore. 

In general this type of shared services of type `Arc<dyn Svc + Send + Sync + 'static>` worked well in other places of our code base (even if it is not the most idiomatic Rust way to do things) and I think it is easy to use.